openmc/deplete/chain.py

```diff
@@ -703,6 +704,65 @@ def form_matrix(self, rates, fission_yields=None):
         # Return CSC representation instead of DOK
         return matrix.tocsc()
 
+    def add_redox_term(self, matrix, buffer, oxidation_states):
+        """Adds a redox term to the depletion matrix from data contained in
+        the matrix itself and a few user-inputs.
+
+        The redox term to add to the buffer nuclide :math:`N_b` can be written
+        as: :math:`\frac{dN_b(t)}{dt} =
+                \cdots + \frac{1}{Ox_b}\sum_i N_i\left( L_{ii}Ox_i -
+                \sum_j G_{i\rightarrow j } Ox_j\right)`
+
+        where :math:`Ox_b` and :math:`Ox_j` are the oxidation states for the
+        corresponding buffer elmenent and j-th nuclide.
+        The first term in the right hand side represent the losses in the
+        diagonal terms of the Bateman matrix, for each nuclide :math:`i`, and
+        the second one the gains in the off-diagonal terms, multiplied by their
+        respective oxidation states.
+
+        Parameters
+        ----------
+        matrix : scipy.sparse.csr_matrix
+            Sparse matrix representing depletion
+        buffer : dict
+            Dictionary of buffer nuclides to be added to keep redox constant,
+            where keys are nuclide names and values fractions to 1.
+        oxidation_states : dict
+            User-defined oxidation states for elements.
+        Returns
+        -------
+        matrix : scipy.sparse.csr_matrix
+            Sparse matrix with redox term added
+        """
+        # Elements list with the same size as self.nuclides
+        elements = [re.split(r'\d+', nuc.name)[0] for nuc in self.nuclides]
+
+        # Match oxidation states with all elements and add 0 if not data
+        ox = np.array([oxidation_states[elm] \
+                        if elm in oxidation_states else 0 for elm in elements])
+
+        # Buffer idx with nuclide index as value
+        buffer_idx = {nuc:self.nuclide_dict[nuc] for nuc in buffer}
+        array = matrix.toarray()
+        redox = np.array([])
+
+        # calculate the redox array
+        for i in range(len(self)):
+            # Gains: multiply each column of the depletion matrix by the oxidation states
+            # vector, excluding the i-th terms
+            gains = np.concatenate((array[:i,i], array[i+1:,i])) * np.delete(ox, i)
+            # Loss: multiply the i-th term by its corresponding oxidation state value
+            loss = array[i,i] * ox[i]
+            # Calculate the redox term
+            redox = np.append(redox, loss + sum(gains))
```

Comment:
I've stared at this a long time and it looks to me like both terms here are positive, which would mean there's really no point of treating the off-diagonal terms separately. Based on the equation, it looks like the gains should be negative?

Reply:
Sounds good, thanks @church89! Just ping me once this PR is ready for another review